utilities/src/serialize/impls.rs

```diff
@@ -150,271 +196,257 @@ impl<T: CanonicalDeserialize> CanonicalDeserialize for Option<T> {
-    }
-}
+// impl<'b, T> Valid for Cow<'b, T>
+// where
+//     T: ToOwned + Sync + Valid + Send,
+//     <T as ToOwned>::Owned: CanonicalDeserialize + Send + Valid,
+// {
+//     #[inline]
+//     fn check(&self) -> Result<(), SerializationError> {
+//         <<T as ToOwned>::Owned>::check(self.as_ref().borrow())
+//     }
+
+//     #[inline]
+
+//     fn batch_check<'a>(batch: impl Iterator<Item = &'a Self>) -> Result<(), SerializationError>
+//     where
+//         Self: 'a
+//     {
+//         <<T as ToOwned>::Owned>::batch_check(batch.map(|v| v.as_ref()))
+//     }
+// }
+
+// impl<'a, T> CanonicalDeserialize for Cow<'a, T>
+// where
+//     T: ToOwned + Valid + Valid + Sync + Send,
+//     <T as ToOwned>::Owned: CanonicalDeserialize + Valid + Send,
+// {
+//     #[inline]
+//     fn deserialize_with_mode<R: Read>(reader: R, compress: Compress, validate: Validate) -> Result<Self, SerializationError> {
+//         Ok(Cow::Owned(<T as ToOwned>::Owned::deserialize_with_mode(reader, compress, validate)?))
+//     }
+// }
 
 impl<T: CanonicalSerialize> CanonicalSerialize for Vec<T> {
     #[inline]
-    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), SerializationError> {
-        self.as_slice().serialize(writer)
+    fn serialize_with_mode<W: Write>(&self, mut writer: W, compress: Compress) -> Result<(), SerializationError> {
+        self.as_slice().serialize_with_mode(&mut writer, compress)
     }
 
     #[inline]
-    fn serialized_size(&self) -> usize {
-        self.as_slice().serialized_size()
+    fn serialized_size(&self, compress: Compress) -> usize {
+        self.as_slice().serialized_size(compress)
     }
+}
 
+impl<T: Valid> Valid for Vec<T> {
     #[inline]
-    fn serialize_uncompressed<W: Write>(&self, writer: &mut W) -> Result<(), SerializationError> {
-        self.as_slice().serialize_uncompressed(writer)
+    fn check(&self) -> Result<(), SerializationError> {
+        T::batch_check(self.iter())
     }
 
     #[inline]
-    fn uncompressed_size(&self) -> usize {
-        self.as_slice().uncompressed_size()
+    fn batch_check<'a>(batch: impl Iterator<Item = &'a Self> + Send) -> Result<(), SerializationError>
+    where
+        Self: 'a,
+    {
+        T::batch_check(batch.flat_map(|v| v.iter()))
     }
 }
 
 impl<T: CanonicalDeserialize> CanonicalDeserialize for Vec<T> {
     #[inline]
-    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, SerializationError> {
-        let len = u64::deserialize(reader)?;
-
-        if mem::size_of::<T>().saturating_mul(len as usize) > 1024 * 1024 * 1024 {
-            return Err(SerializationError::InvalidData);
-        }
-
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        compress: Compress,
+        validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let len = u64::deserialize_with_mode(&mut reader, compress, validate)?;
         let mut values = Vec::with_capacity(len as usize);
```

Comment:
```suggestion
        let mut values = Vec::new();
```
the same rationale as in https://github.com/AleoHQ/snarkVM/pull/791/files#r872152254